Random connection failure in readInitialHandshake()

Hello there,

First of all, thank you for your work!
I'm currently using the latest commit of the driver (https://github.com/go-mysql-org/go-mysql/commit/eb2c6d100ad80877ec2b88026cad8cf005442f51) with Go 1.22.2, and I get an error randomly when I try to open a database connection. 
4 times over 5, it works, but sometimes it ends up with this error: `runtime error: slice bounds out of range [:9] with capacity 8`.
The MySQL server version is: `8.0.37`. I'm using MySQL in a docker container (`mysql:8.0` image).

I open the connection in a very old-fashion way:
```
	dsn := fmt.Sprintf("%s:%s@%s/%s?ssl=true", mySqlUsername, mySqlPassword, mySqlHost, mySqlSchema)
	db, err := sql.Open("mysql", dsn)
	if err != nil {
		t.Fatalf("failed to connect to mysql: %s", err)
	}
	defer db.Close()
```
and do a simple `select count(1)` query.

And here is the stacktrace I have 1 time over 5:
```
panic: runtime error: slice bounds out of range [:9] with capacity 8

goroutine 1542 [running]:
github.com/go-mysql-org/go-mysql/client.(*Conn).readInitialHandshake(0xc002bb8c30)
	/home/runner/go/pkg/mod/github.com/go-mysql-org/go-mysql@v1.7.1-0.20240508103851-eb2c6d100ad8/client/auth.go:57 +0x8c6
github.com/go-mysql-org/go-mysql/client.(*Conn).handshake(0xc002bb8c30)
	/home/runner/go/pkg/mod/github.com/go-mysql-org/go-mysql@v1.7.1-0.20240508103851-eb2c6d100ad8/client/conn.go:157 +0x1c
github.com/go-mysql-org/go-mysql/client.ConnectWithDialer({0xcad3f0, 0xc0002c3340}, {0x0, 0x0}, {0xc002cc53dd, 0xf}, {0xc002cc53c8, 0x7}, {0xc002cc53d0, 0xc}, ...)
	/home/runner/go/pkg/mod/github.com/go-mysql-org/go-mysql@v1.7.1-0.20240508103851-eb2c6d100ad8/client/conn.go:127 +0x545
github.com/go-mysql-org/go-mysql/client.ConnectWithContext({0xcad3f0?, 0xc0002c3340?}, {0xc002cc53dd?, 0x1?}, {0xc002cc53c8?, 0xf?}, {0xc002cc53d0?, 0x7?}, {0xc002cc53ed, 0x6}, ...)
	/home/runner/go/pkg/mod/github.com/go-mysql-org/go-mysql@v1.7.1-0.20240508103851-eb2c6d100ad8/client/conn.go:79 +0x105
github.com/go-mysql-org/go-mysql/client.Connect({0xc002cc53dd, 0xf}, {0xc002cc53c8, 0x7}, {0xc002cc53d0, 0xc}, {0xc002cc53ed, 0x6}, {0xc000051948, 0x1, ...})
	/home/runner/go/pkg/mod/github.com/go-mysql-org/go-mysql@v1.7.1-0.20240508103851-eb2c6d100ad8/client/conn.go:73 +0x11a
github.com/go-mysql-org/go-mysql/driver.driver.Open({}, {0xc002cc5380?, 0xc002e10001?})
	/home/runner/go/pkg/mod/github.com/go-mysql-org/go-mysql@v1.7.1-0.20240508103851-eb2c6d100ad8/driver/driver.go:102 +0x19b
database/sql.dsnConnector.Connect(...)
	/opt/hostedtoolcache/go/1.22.2/x64/src/database/sql/sql.go:791
database/sql.(*DB).conn(0xc002bb8b60, {0xcad2a0, 0x114e4c0}, 0x1)
	/opt/hostedtoolcache/go/1.22.2/x64/src/database/sql/sql.go:1415 +0x71e
database/sql.(*DB).query(0xc002bb8b60, {0xcad2a0, 0x114e4c0}, {0xb8f886, 0x10}, {0x0, 0x0, 0x0}, 0xa0?)
	/opt/hostedtoolcache/go/1.22.2/x64/src/database/sql/sql.go:1749 +0x57
database/sql.(*DB).QueryContext.func1(0xa0?)
	/opt/hostedtoolcache/go/1.22.2/x64/src/database/sql/sql.go:1732 +0x4f
database/sql.(*DB).retry(0xc000051d28?, 0xc000051cd0)
	/opt/hostedtoolcache/go/1.22.2/x64/src/database/sql/sql.go:1566 +0x42
database/sql.(*DB).QueryContext(0xb88d5d?, {0xcad2a0?, 0x114e4c0?}, {0xb8f886?, 0x3?}, {0x0?, 0x41097b?, 0x7f3a2a14c5b8?})
	/opt/hostedtoolcache/go/1.22.2/x64/src/database/sql/sql.go:1731 +0xc5
database/sql.(*DB).Query(...)
	/opt/hostedtoolcache/go/1.22.2/x64/src/database/sql/sql.go:1745
```

I managed to debug the `readInitialHandshake()` and capture the `data` value which cause this issue, and `data` contains only this: `<0x0b><0x08><0x05><0x1a><0x00>`:
![image](https://github.com/go-mysql-org/go-mysql/assets/4442140/a2bc7a97-eeef-4c7f-bf60-93e32f40aeec)


Any clue?
I don't know if it's related to recent changes. Let me know if I can add some flags or verbosity to have much information to share.

Thanks!

--- a/client/auth.go
+++ b/client/auth.go
@@ -42,8 +42,14 @@ func (c *Conn) readInitialHandshake() error {
 		return errors.Annotate(c.handleErrorPacket(data), "read initial handshake error")
 	}
 
-	if data[0] < MinProtocolVersion {
-		return errors.Errorf("invalid protocol version %d, must >= 10", data[0])
+	if data[0] != ClassicProtocolVersion {
+		if data[0] == XProtocolVersion {
+			return errors.Errorf(
+				"invalid protocol version %d, expected 10. "+
+					"This might be X Protocol, make sure to connect to the right port",
+				data[0])
+		}
+		return errors.Errorf("invalid protocol version %d, expected 10", data[0])
 	}
 	pos := 1
 

--- a/mysql/const.go
+++ b/mysql/const.go
@@ -1,9 +1,10 @@
 package mysql
 
 const (
-	MinProtocolVersion byte   = 10
-	MaxPayloadLen      int    = 1<<24 - 1
-	TimeFormat         string = "2006-01-02 15:04:05"
+	ClassicProtocolVersion byte   = 10
+	XProtocolVersion       byte   = 11
+	MaxPayloadLen          int    = 1<<24 - 1
+	TimeFormat             string = "2006-01-02 15:04:05"
 )
 
 const (

